ui: Tests on react-routing components

koorosh opened the pull request:
Depends on PR: https://github.com/cockroachdb/cockroach/pull/43846
`yarn-vendor` PR has to be merged: https://github.com/cockroachdb/yarn-vendored/pull/15

Current PR includes minimal changes to functionality (only those which are required to isolate or get
access to individual components under tests):
- `src/index.tsx` file has been refactored into two parts: rendering to DOM and independent `App` component;
-  <Helmet> component (`react-helmet` library) has been refactored to avoid known issue during tests execution (https://github.com/nfl/react-helmet/issues/373)
- old version of react-router and `connect` component in tandem do not work well because connected
component aren't valid react components. As result validation for correct routing was done by re-exporting wrapped components (instead of connected). It allowed to simplify testing with `enzyme#find` function which requires valid react component as an argument;
- Tests are grouped in suites per route to maintain related tests per route;
- There is two kinds of tests which test correct rendering of component for
specific test and another test which validates correct redirection for routes;
- Tests can be refactored to be constructed from configuration to reduce repeated
tests, but it will be actual after migrating to latest version of react-router lib.

koorosh committed:
ui: Upgrade React version and react-dependent libraries

With current changes, react is upgraded to latest version (16.12.0) and
following libraries are upgraded as well: react-redux, redux, redux-saga.
These upgrades of major versions has required to do some code adjustments
to properly migrate to newest versions.
- redux. Provide valid type arguments for Dispatch type.
- redux-saga. Changed import and usage of 'delay' function.
- react-redux. `connect` function failed to correctly resolve returned type
for `mapDispatchToProps` argument. As result it is provided as a function
instead of object argument. This hack allows implicitly resolve types.
- `CachedDataReducer` class is extended with one more optional type argument
to specify the list of allowed `actionNamespace` literals instead of `string`.
It was necessary to have strictly defined CachedDataReducer interface.
- `normalizeConnectedComponent` function is used to do simple mapping from
react's ExoticComponent to valid react component. This resolves issue of
providing ConnectedComponents (which are ExoticComponent's, not regular react
components) to Route component which has strict type validation of component types.

Release note: None

koorosh committed:
ui: Fix version of enzyme-react-adapter library

Release note: None

koorosh committed:
ui: Adjust code structure for testing routes

- Decoupled App from rendering app into DOM.
- Defined tests structure and initial tests.

Release note: None

koorosh committed:
ui: Fix warning messages during karma tests execution

- Karma produced warning messages related to compilation of .ts|.tsx files to .js
due to incorrect webpack configuration.
- Added macha reporter to get readable test reports.

Release note: None

koorosh committed:
ui: Add tests for /overview, /metrics, and /database(s) routes

- Tests are grouped in suites per route to maintain related tests per route;
- There is two kinds of tests which test correct rendering of component for
specific test and another test which validates correct redirection for routes;
- Tests can be refactored to be constructed from configuration to reduce repeated
tests, but it will be actual after migrating to latest version of react-router lib.
- Added additional exports for components which is wrapped by Connect HOC because
connect doesn't return valid react component, as result `enzym` library cannot find
wrapped component in mounted components tree. This can be changed back with latest
version of `react-router` + `react-connected-router` library which don't have this issue.

Release note: None

koorosh committed:
ui: Fix linting errors

Release note: None

koorosh committed:
ui: Refactor alertDateSync subscription out of App component

Change back subscription to alertDataSync out of App context so
it doesn't break existing behaviour.

Release note: None

koorosh committed:
ui: Clean up resources after tests complete

Release note: None

koorosh committed:
ui: Add tests for /statement(s) routes

Release note: None

koorosh committed:
ui: Add tests for /debug/* routes

Release note: None

koorosh committed:
ui: Add tests for /raft/* routes

Release note: None

koorosh committed:
ui: Add tests for /cluster/* routes

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44078)
<!-- Reviewable:end -->

koorosh committed:
ui: Resolve conflicts after rebasing against master

- tests command is disabled for now because current state doesn't
contain all fixes to run tests without failure.
- Decommissioned Node History page was added so code has to be adjusted.

Release note: None

koorosh committed:
ui: Resolve conflicts after rebasing against master branch

- Wrap NodeCanvasContainer and statementsPage components with `bindActionCreators`
- Add route for DecommissionedNodeHistory page

Release note: None

koorosh committed:
ui: Fix linting error

Release note: None

koorosh committed:
ui: Fix conflicts after rebasing against 'ui-react-version-upgrade' branch

Release note: None

koorosh committed:
ui: Export Connected and pure Component classes for correct unit testing

Testing with Enzyme framework is straight-froward and easy, but there is once
gotcha: ConnectedComponents (wrapped with Connect function) are not recognized
by Enzyme as react components and it is not possible to use enzymeWrapper.find(Component)
helper function to check whether connected component is rendered in DOM or not.

To make it clear there is an example how Connected Component is recognized in dom:
```
<[object Object] location={{...}} params={{...}} route={{...}} ...
```
as [object] not a Component.

To overcome this problem "not connected" child components are exported in
addition to connected. It allows to write tests against these components.

Release note: None

koorosh committed:
ui: Get rid of useless function wrappers

It is not necessary to wrap action creator object
with function when connecting it to component.

Release note: None

koorosh committed:
ui: Fix missed routes for Statements and Decommissioned node list pages

- Remove unnecessary re-exporting from index.ts file
- Simplify exporting of DecommissionedNodeList component

Release note: None

koorosh committed:
ui: Add tests on routing to /reports views

Release note: None

dhartunian commented:
bors r+
